accounting/src/delta/combine.rs

```diff
@@ -15,30 +15,41 @@
 
 use common::primitives::{signed_amount::SignedAmount, Amount};
 
-use crate::{
-    delta::delta_data_collection::{DataDelta, DeltaMapElement},
-    error::Error,
-};
+use crate::{delta::delta_data_collection::DeltaMapElement, error::Error};
 
 /// Combine data with an element of `DeltaDataCollection`.
 /// An element can be either a Delta or a result of delta undo.
-pub fn combine_data_with_delta<T: Clone>(
+pub fn combine_data_with_delta<T: Clone + PartialEq>(
     lhs: Option<&T>,
     rhs: Option<&DeltaMapElement<T>>,
 ) -> Result<Option<T>, Error> {
     match (lhs, rhs) {
         (None, None) => Ok(None),
-        (None, Some(d)) => match d.get_data_delta() {
-            DataDelta::Create(d) => Ok(Some(d.clone())),
-            DataDelta::Modify(_, _) => Err(Error::ModifyNonexistingData),
-            DataDelta::Delete(_) => Err(Error::RemoveNonexistingData),
-        },
+        (None, Some(d)) => {
+            let delta = d.get_data_delta();
+            match (&delta.prev(), &delta.next()) {
+                (None, None) => Ok(None),
+                (None, Some(d)) => Ok(Some(d.clone())),
+                (Some(_), None) => Err(Error::RemoveNonexistingData),
+                (Some(_), Some(_)) => Err(Error::ModifyNonexistingData),
+            }
+        }
         (Some(p), None) => Ok(Some(p.clone())),
-        (Some(_), Some(d)) => match d.get_data_delta() {
-            DataDelta::Create(_) => Err(Error::DataCreatedMultipleTimes),
-            DataDelta::Modify(_, d) => Ok(Some(d.clone())),
-            DataDelta::Delete(_) => Ok(None),
-        },
+        (Some(data), Some(delta)) => {
+            let delta = delta.get_data_delta();
+            match (&delta.prev(), &delta.next()) {
+                (None, None) => Err(Error::RemoveNonexistingData),
+                (None, Some(_)) => Err(Error::DataCreatedMultipleTimes),
```

Comment:
My intention was to treat these cases more user-friendly. For example, if Data is `Some` and Delta is `(None, Some)` that means that the data was already created and we try to create it once more.

Reply:
In the first case, for example, the data do exist (look at the `match` arm one level up) but the error says `RemoveNonexistingData`. This might be fixed in the more recent version. If so, please ignore this comment.